src/splitwise.js

```diff
@@ -614,28 +627,79 @@ class Splitwise {
       logger({ level: LOG_LEVELS.ERROR, message });
       throw new Error(message);
     }
+    // check if no redirect url supplied to useOauth2
+    if (useOauth2 && !redirect_uri) {
+      const message = 'Redirect url required for OAuth2';
+      logger({ level: LOG_LEVELS.ERROR, message });
+      throw new Error(message);
+    }
 
     const oauth2 = new OAuth2(
       consumerKey,
       consumerSecret,
-      'https://secure.splitwise.com/',
-      null,
-      'oauth/token',
+      SPLITWISE_ENDPOINTS.baseUrl,
+      useOauth2 ? SPLITWISE_ENDPOINTS.authorizeUrl : null,
+      SPLITWISE_ENDPOINTS.accessTokenUrl,
       null
     );
 
-    const accessTokenPromise = (() => {
+    const generateState = () => {
+      this.state = crypto.randomBytes(20).toString('hex');
+      return this.state;
+    }
+
+    this.getAuthorizationUrl = () => {
+      if (!useOauth2) return "";
+      return oauth2.getAuthorizeUrl({
+        redirect_uri,
+        scope: '',
+        state: generateState(),
+        response_type: 'code'
+      });
+    }
+
+    const verifyState = state => {
+      if (!useOauth2) return true;
+      return state === this.state;
+    }
+
+    const getAccessTokenFromAuthCode = () => {
+      return new Promise((resolve, reject) => {
+        if(!this.authCode) 
+          return reject(`No auth code generated yet. Visit ${this.getAuthorizationUrl()} and login. You need a callback url in your project as mentioned in callback url in your registered splitwise app. Then call \`getAccessToken()\` to register the access token.`);
+        if(this.accessToken) return resolve(this.accessToken);
+        return oauth2.getOAuthAccessToken(this.authCode, {
+          code: this.authCode,
+          redirect_uri,
+          grant_type: 'authorization_code'
+        }, (err, accessToken) => {
+          if (err) {
+            return reject(err);
+          }
+          oauth2.useAuthorizationHeaderforGET(true);
+          this.accessToken = accessToken;
+          return resolve(true);
+        })
+      })
+    }
+
+
+    const accessTokenPromise = () => {
       if (accessToken) {
         logger({ message: 'using provided access token' });
         return Promise.resolve(accessToken);
       }
       logger({ message: 'making request for access token' });
       return getAccessTokenPromise(logger, oauth2);
-    })();
+    };
+    if (!useOauth2) {
+      // earlier an IIFE. But now, this is called only in case of client_credentials
+      accessTokenPromise();
+    }
```

Comment:
I think this code would really benefit from having a clear separation between the client credentials code and the authorization code code. maybe something like a switch statement over the new `grantType` argument which contains the relevant pieces.

Reply:
Called by default while initializing so as to acquire a token right away. This is just a single case. We cant get the token right away in case of authorization_code